### ISSUE TITLE
Allow DropdownButton to flex to a maximum width of 100%

### ISSUE THREAD

**Issue by oodavid** [flutter/flutter#88572]
## Proposal

Allow `DropdownButton` to behave in two new ways:

1 - Automatically expand to a maximum width of 100%
2 - Let the width be based on the selected value *not* the maximum width of the `items`

I have an example _in HTML and CSS_ that demonstrates the flex behaviour. I've yet to figure it out with Flutter:

https://codepen.io/oodavid/pen/OJmKwzX?editors=1100

A short video explaining it:

https://www.loom.com/share/3876987004fc4a41a2a9d170b70876c1

**Comment by maheshmnj** [comment]
Hi @oodavid, thanks for filing the issue.
 I am not sure I understand the purpose of the first proposal, I do see that the dropdown does take 100% of the width with `isExpanded: true`,  setting it to false (default) it would just take the max of its children's width. I am not sure why would we want to allow dropdown to take 100% of its width by setting `isExpanded: false`wouldn't that beat the purpose of having an isExpanded parameter?

**Comment by oodavid** [comment]
Hi @maheshmnj,

We're trying to adhere to the idea that fields should be an appropriate size for their expected input, ie:

![1_ZRRH7Tq4BClBucqf4ElpBA](https://user-images.githubusercontent.com/46879/130454171-17fe27b6-ef8b-4f73-81ba-da5057e70be7.png)

<table>
  <tr>
    <th><code>isExpanded: false</code></th>
    <th><code>isExpanded: true</code></th>
  </tr>
  <tr>
    <td>
      ✅ The first has short values, and renders small<br>
      ❌ The second has one long value, causing an overflow error
    </td>
    <td>
      ✅ The second input no longer has an overflow error<br>
      ❌ But now the first input is huge
    </td>
  </tr>
  </tr>
  <tr>
    <td><img src="https://user-images.githubusercontent.com/46879/130452392-bc11e819-c5a6-4b3c-81d8-496920d070d0.gif"></td>
    <td><img src="https://user-images.githubusercontent.com/46879/130452384-16729cf3-7391-41a4-bb9c-120084383abb.gif"></td>
  </tr>
</table>

To recap:

1. Automatically expand to a maximum width of 100%
   Stay small, but don't expand to infinity.
   Don't cause an overflow issue
2. Let the width be based on the selected value not the maximum width of the items
   More nuanced; we'd prefer to have a minimum width, and use the selected value to dictate the size.
   This would mean that the Widget could change size when the value changes.

Happy to add more detail if needed.

**Comment by maheshmnj** [comment]
Thanks for the clarification. This does look like a valid proposal.

**Comment by oodavid** [comment]
Mockup:

<img width="362" alt="Screenshot 2021-08-23 at 16 40 08" src="https://user-images.githubusercontent.com/46879/130476268-b5210c6b-2e9d-4890-b30b-eb56701897d7.png">

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Padding(
          padding: EdgeInsets.all(8.0),
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              MyDropdown(value: 'Short'),
              SizedBox(height: 16),
              MyDropdown(value: 'This has a medium value'),
              SizedBox(height: 16),
              MyDropdown(
                value:
                    'Latin is Long. Do you like latin? Sic semper tyrannis! Aut cum scuto aut in scuto. Lorem ipsum dolor sit amet',
              ),
              SizedBox(height: 16),
              MyDropdown(
                value:
                    'Latin is Long. Do you like latin? Sic semper tyrannis! Aut cum scuto aut in scuto. Lorem ipsum dolor sit amet',
                overflow: TextOverflow.ellipsis,
              ),
            ],
          ),
        ),
      ),
    );
  }
}

class MyDropdown extends StatelessWidget {
  const MyDropdown({
    Key? key,
    required this.value,
    this.overflow,
  }) : super(key: key);

  final String value;
  final TextOverflow? overflow;

  @override
  Widget build(BuildContext context) {
    return Container(
      decoration: BoxDecoration(
        border: Border.all(
          color: Colors.grey,
        ),
        color: Colors.white,
        borderRadius: BorderRadius.all(
          Radius.circular(4),
        ),
      ),
      child: Row(
        mainAxisSize: MainAxisSize.min,
        children: <Widget>[
          Flexible(
            child: Container(
              padding: EdgeInsets.symmetric(
                horizontal: 8,
                vertical: 4,
              ),
              child: Text(
                value,
                overflow: overflow,
              ),
            ),
          ),
          Container(
            padding: EdgeInsets.symmetric(
              horizontal: 4,
            ),
            child: Icon(Icons.arrow_drop_down),
          ),
        ],
      ),
    );
  }
}
```

**Comment by oodavid** [comment]
I think the fix is straightforward, but would introduce deprecations to the API*

Ultimately the Widget handles `isExpanded` like so:

https://github.com/flutter/flutter/blob/02b46b0d775166463416b85c80ad5c428ad8be6f/packages/flutter/lib/src/material/dropdown.dart#L1465-L1468

Seems trivial to replace `bool isExpanded` with `FlexFit fit`, then replace the above with

```dart
Flexible(child: innerItemsWidget, fit: fit)
```

_* NB: I'm not sure how you handle API changes with pull-requests. I'm happy to read any docs about that before making a pull-request._

**Comment by HansMuller** [comment]
CC @QuncCccccc
